keps/sig-autoscaling/20191111-autoscaling-with-multiple-thresholds.md

```diff
@@ -0,0 +1,202 @@
+---
+title: Multi thresholds based Autoscaling.
+authors:
+  - "@CharlyF"
+owning-sig: sig-autoscaling
+participating-sigs:
+reviewers:
+  - TBD
+approvers:
+  - TBD
+editor: TBD
+creation-date: 2019-11-11
+last-updated: 2019-11-11
+status: provisional
+see-also:
+  - "/keps/sig-autoscaling/20190307-configurable-scale-velocity-for-hpa.md"
+replaces:
+superseded-by:
+---
+
+# Multi thresholds based Autoscaling
+
+## Table of Contents
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+- [Proposal](#proposal)
+  - [User Stories](#user-stories-optional)
+    - [Canonical use case of the bounds to control the primary scaling condition.](#Canonical-use-case-of-the-bounds-to-control-the-primary-scaling-condition)
+    - [Restriction of the scaling events.](#Restriction-of-the-scaling-events)
+  - [Implementation Details/Notes/Constraints](#implementation-detailsnotesconstraints)
+    - [The Pseudo Code](#the-pseudo-code)
+- [Design Details](#design-details)
+  - [Upgrade / Downgrade Strategy](#upgrade--downgrade-strategy)
+<!-- /toc -->
+
+## Summary
+
+This is a proposal to extend the Horizontal Pod Autoscaler (or HPA) Controller to allow for a better configurability of the scaling behavior.
+Namely, avoiding the sawtooth pattern by introducing two thresholds as low and high bounds.
+We (Datadog) have been using a Custom Resource internally, the Watermark Pod Autoscaler or WPA, and we believe the community could benefit from this feature. 
+Feel free to check out [the official repository](https://github.com/DataDog/watermarkpodautoscaler) and give us your feedback.
+
+Worth mentioning, we have also included some of the features introduced by @gliush
+and @arjunrn in in the [Configurable Scale Velocity KEP](https://github.com/kubernetes/enhancements/blob/master/keps/sig-autoscaling/20190307-configurable-scale-velocity-for-hpa.md) in our controller as we believe those features are must-haves for the next verison of the HPA Controller.
+
+
+## Motivation
+
+- Set a high and a low bound at the resource level to better control autoscaling events.
+
+### Goals
+
+The goal is to improve the ability to customize the parameters triggering a scaling event. We identified 4 vectors of improvement:
+
+- Specifying multiple thresholds on a per HPA basis.
+- Modifying the tolerance (`--horizontal-pod-autoscaler-tolerance`) at the HPA level.
```

Comment:
Toleration seems to give the same functionality as low and high watermarks/thresholds/targets but in much cleaner way. We already have too many "targets" in HPA and adding couple more won't help. 
Unless you have a use case where low/high watermark works and tolerance doesn't please consider removing watermarks from this proposal.